# From Raw Data to Predictions

This notebook is designed as a follow-up to the [Machine Learning with the Experts: School Budgets](https://www.datacamp.com/courses/machine-learning-with-the-experts-school-budgets) course on Datacamp. We won't explain all the tools and techniques we use here. If you're curious about any of the tools, code, or methods used here, make sure to check out the course!

In [1]:
%matplotlib inline
from __future__ import division
from __future__ import print_function

import matplotlib.pyplot as plt
import numpy as np
import pandas as pd

import os
import sys

# add the 'src' directory as one where we can import modules
src_dir = os.path.join(os.getcwd(), os.pardir, 'src')
sys.path.append(src_dir)

from data.multilabel import multilabel_sample_dataframe, multilabel_train_test_split
from features.SparseInteractions import SparseInteractions
from models.metrics import multi_multi_log_loss

# Load Data

First, we'll load the entire trainging data set available from DrivenData.


In [12]:
path_to_training_data = os.path.join(os.getcwd(),os.pardir,
                                     'data','raw',
                                     'TrainingSet.csv')

df = pd.read_csv(path_to_training_data, index_col=0)

print(df.shape)

(400277, 25)


In [13]:
df.head()

Function          Use          Sharing Reporting  \
134338     Teacher Compensation  Instruction  School Reported    School   
206341                 NO_LABEL     NO_LABEL         NO_LABEL  NO_LABEL   
326408     Teacher Compensation  Instruction  School Reported    School   
364634  Substitute Compensation  Instruction  School Reported    School   
47683   Substitute Compensation  Instruction  School Reported    School   

       Student_Type Position_Type               Object_Type     Pre_K  \
134338     NO_LABEL       Teacher                  NO_LABEL  NO_LABEL   
206341     NO_LABEL      NO_LABEL                  NO_LABEL  NO_LABEL   
326408  Unspecified       Teacher  Base Salary/Compensation  Non PreK   
364634  Unspecified    Substitute                  Benefits  NO_LABEL   
47683   Unspecified       Teacher   Substitute Compensation  NO_LABEL   

         Operating_Status            Object_Description  \
134338  PreK-12 Operating                           NaN   
206341      Non-Operating           CONTRACTOR SERVICES   
326408  PreK-12 Operating  Personal Services - Teachers   
364634  PreK-12 Operating             EMPLOYEE BENEFITS   
47683   PreK-12 Operating  TEACHER COVERAGE FOR TEACHER   

                    ...               Sub_Object_Description  \
134338              ...                                  NaN   
206341              ...                                  NaN   
326408              ...                                  NaN   
364634              ...                                  NaN   
47683               ...                                  NaN   

       Location_Description  FTE     Function_Description  \
134338                  NaN  1.0                      NaN   
206341                  NaN  NaN                 RGN  GOB   
326408                  NaN  1.0                      NaN   
364634                  NaN  NaN  UNALLOC BUDGETS/SCHOOLS   
47683                   NaN  NaN              NON-PROJECT   

       Facility_or_Department              Position_Extra      Total  \
134338                    NaN               KINDERGARTEN   50471.810   
206341                    NaN                UNDESIGNATED   3477.860   
326408                    NaN                     TEACHER  62237.130   
364634                    NaN  PROFESSIONAL-INSTRUCTIONAL     22.300   
47683                     NaN  PROFESSIONAL-INSTRUCTIONAL     54.166   

                   Program_Description        Fund_Description  \
134338                    KINDERGARTEN            General Fund   
206341   BUILDING IMPROVEMENT SERVICES                     NaN   
326408           Instruction - Regular  General Purpose School   
364634  GENERAL MIDDLE/JUNIOR HIGH SCH                     NaN   
47683    GENERAL HIGH SCHOOL EDUCATION                     NaN   

                               Text_1  
134338                            NaN  
206341  BUILDING IMPROVEMENT SERVICES  
326408                            NaN  
364634            REGULAR INSTRUCTION  
47683             REGULAR INSTRUCTION  

[5 rows x 25 columns]

# Resample Data

400,277 rows is too many to work with locally while we develop our approach. We'll sample down to 10,000 rows so that it is easy and quick to run our analysis.

We'll also create dummy variables for our labels and split our sampled dataset into a training set and a test set.

In [51]:
LABELS = ['Function',
          'Use',
          'Sharing',
          'Reporting',
          'Student_Type',
          'Position_Type',
          'Object_Type', 
          'Pre_K',
          'Operating_Status']

NON_LABELS = [c for c in df.columns if c not in LABELS]

SAMPLE_SIZE = 5000

sampling = multilabel_sample_dataframe(df,
                                       pd.get_dummies(df[LABELS]),
                                       size=SAMPLE_SIZE,
                                       min_count=25,
                                       seed=43)

dummy_labels = pd.get_dummies(sampling[LABELS])

X_train, X_test, y_train, y_test = multilabel_train_test_split(sampling[NON_LABELS],
                                                               dummy_labels,
                                                               0.2,
                                                               min_count=3,
                                                               seed=43)

In [58]:
X_train.shape

(4000, 16)

# Create preprocessing tools

We need tools to preprocess our text and numeric data. We'll create those tools here. The `combine_text_columns` function will take a DataFrame of text columns and return a single series where all of the text in the columns has been joined together.

We'll then create `FunctionTransformer` objects that select our text and numeric data from the dataframe.

Finally, we create a custom scoring method that uses the `multi_multi_log_loss` function that is the evaluation metric for the competition.

In [34]:
NUMERIC_COLUMNS = ['FTE', "Total"]

def combine_text_columns(data_frame, to_drop=NUMERIC_COLUMNS + LABELS):
    """ Takes the dataset as read in, drops the non-feature, non-text columns and
        then combines all of the text columns into a single vector that has all of
        the text for a row.
        
        :param data_frame: The data as read in with read_csv (no preprocessing necessary)
        :param to_drop (optional): Removes the numeric and label columns by default.
    """
    # drop non-text columns that are in the df
    to_drop = set(to_drop) & set(data_frame.columns.tolist())
    text_data = data_frame.drop(to_drop, axis=1)
    
    # replace nans with blanks
    text_data.fillna("", inplace=True)
    
    # joins all of the text items in a row (axis=1)
    # with a space in between
    return text_data.apply(lambda x: " ".join(x), axis=1)


In [52]:
from sklearn.preprocessing import FunctionTransformer

get_text_data = FunctionTransformer(combine_text_columns, validate=False)
get_numeric_data = FunctionTransformer(lambda x: x[NUMERIC_COLUMNS], validate=False)

In [59]:
get_text_data.fit_transform(sampling.head(5))

38     OTHER PURCHASED SERVICES  SCHOOL-WIDE SCHOOL P...
70     Extra Duty Pay/Overtime For Support Personnel ...
198    Supplemental *  Operation and Maintenance of P...
209    REPAIR AND MAINTENANCE SERVICES  PUPIL TRANSPO...
614     GENERAL EDUCATION LOCAL EDUCATIONAL AIDE,70 H...
dtype: object

In [60]:
get_numeric_data.fit_transform(sampling.head(5))

FTE         Total
38    NaN    653.460000
70    NaN   2153.530000
198   NaN  -8291.860000
209   NaN    618.290000
614  0.71  21747.666875

In [38]:
from sklearn.metrics.scorer import make_scorer

log_loss_scorer = make_scorer(multi_multi_log_loss)

# Train model pipeline

Now we'll train the final pipeline from the course that takes text and numeric data, does the necessary preprocessing, and trains the classifier.

In [39]:
from sklearn.feature_selection import chi2, SelectKBest

from sklearn.pipeline import Pipeline, FeatureUnion

from sklearn.preprocessing import Imputer
from sklearn.feature_extraction.text import HashingVectorizer

from sklearn.multiclass import OneVsRestClassifier
from sklearn.linear_model import LogisticRegression

from sklearn.preprocessing import MaxAbsScaler

TOKENS_ALPHANUMERIC = '[A-Za-z0-9]+(?=\\s+)'

In [63]:
%%time

# set a reasonable number of features before adding interactions
chi_k = 10

# create the pipeline object
pl = Pipeline([
        ('union', FeatureUnion(
            transformer_list = [
                ('numeric_features', Pipeline([
                    ('selector', get_numeric_data),
                    ('imputer', Imputer())
                ])),
                ('text_features', Pipeline([
                    ('selector', get_text_data),
                    ('vectorizer', HashingVectorizer(token_pattern=TOKENS_ALPHANUMERIC,
                                                     non_negative=True, norm=None, binary=False,
                                                     ngram_range=(1, 2))),
                    ('dim_red', SelectKBest(chi2, chi_k))
                ]))
             ]
        )),
        ('int', SparseInteractions(degree=2)),
        ('scale', MaxAbsScaler()),
        ('clf', OneVsRestClassifier(LogisticRegression()))
    ])

# fit the pipeline to our training data
#pl.fit(X_train, y_train.values)

# print the score of our trained pipeline on our test set
#print("Logloss score of trained pipeline: ", log_loss_scorer(pl, X_test, y_test.values))

Wall time: 1e+03 µs


In [71]:
pl.named_steps

{'union': FeatureUnion(n_jobs=1,
        transformer_list=[('numeric_features', Pipeline(memory=None,
      steps=[('selector', FunctionTransformer(accept_sparse=False,
           func=<function <lambda> at 0x10C21A08>, inv_kw_args=None,
           inverse_func=None, kw_args=None, pass_y='deprecated',
           validate=False)), ('imputer', I...s+)', tokenizer=None)), ('dim_red', SelectKBest(k=10, score_func=<function chi2 at 0x10334DF8>))]))],
        transformer_weights=None),
 'int': SparseInteractions(degree=2, feature_name_separator='_'),
 'scale': MaxAbsScaler(copy=True),
 'clf': OneVsRestClassifier(estimator=LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
           intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
           penalty='l2', random_state=None, solver='liblinear', tol=0.0001,
           verbose=0, warm_start=False),
           n_jobs=1)}

# Predict holdout set and write submission

Finally, we want to use our trained pipeline to predict the holdout dataset. We will write our predictions to a file, `predictions.csv`, that we can submit on [DrivenData](http://www.drivendata.org)!

In [11]:
path_to_holdout_data = os.path.join(os.pardir,
                                    'data',
                                    'TestSet.csv')

# Load holdout data
holdout = pd.read_csv(path_to_holdout_data, index_col=0)

# Make predictions
predictions = pl.predict_proba(holdout)

# Format correctly in new DataFrame: prediction_df
prediction_df = pd.DataFrame(columns=pd.get_dummies(df[LABELS]).columns,
                             index=holdout.index,
                             data=predictions)


# Save prediction_df to csv called "predictions.csv"
prediction_df.to_csv("predictions.csv")

/Users/bull/anaconda/envs/box-plots-sklearn/lib/python3.6/site-packages/IPython/core/interactiveshell.py:2717: DtypeWarning: Columns (5,11) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)
/Users/bull/anaconda/envs/box-plots-sklearn/lib/python3.6/site-packages/sklearn/linear_model/base.py:352: RuntimeWarning: overflow encountered in exp
  np.exp(prob, prob)
